### ISSUE TITLE
Visual timing on macOS

### ISSUE THREAD

**Issue by hoechenberger** [psychopy/psychopy#2250]
@peircej mentioned that he observed a changed timing behavior on Macs since High Sierra, potentially related to triple buffering. I just ran a quick test on my brand-new MacBook Pro 2018 with Touch Bar, running Mojave 10.14.3, and cannot reproduce this issue (at least not when only using the built-in display; cannot hook up an external monitor for now).

```python
from psychopy.visual.window import Window
from psychopy.clock import Clock

clock = Clock()
win = Window(fullscr=True, size=(1440, 900))
n_flips = 10

win.callOnFlip(clock.reset)  # Reset to zero on first flip.

for _ in range(n_flips + 1):
    win.flip()

t = clock.getTime()
win.close()

print(f'Total duration of {n_flips} flips: {round(t*1000, 2)} ms')
```

Output:
```python
/Users/hoechenberger/miniconda3/envs/psychopy-test/bin/python /Users/hoechenberger/Library/Preferences/PyCharm2018.3/scratches/scratch_1.py
pygame 1.9.4
Hello from the pygame community. https://www.pygame.org/contribute.html
0.9357 	WARNING 	Monitor specification not found. Creating a temporary one...
Total duration of 10 flips: 166.5 ms

Process finished with exit code 0
```

I consistently get values around 166–168 ms, which is about the expected duration on a 60 Hz screen.

Running PsychoPy `master`, pyglet 1.3.2.

**Comment by peircej** [comment]
You won't be able to detect it using PsychoPy's reported times:

- the effect is mostly about a 1-frame lag (so frame durations will be fine and stimulus duration should be fine)
- as a lag, this is relative to other events, like labjack commands, keyboard polling, sound outputs (although those also carry their own lag!). To measure it you need to output a pulse on flip and compare with physical stimulus appearance (photodiode)

I think @dvbridges and I don't yet know how much impact this is having on variability.

**Comment by hoechenberger** [comment]
Oh, of course, stupid me! 🥳

Idea: have you tried DISABLING double-buffering on an affected machine? Maybe a double buffer is the default now, and when pyglet requests yet another one, we end up triple-buffering?

--

**Comment by m-macaskill** [comment]
Does this reflect how triple buffering actually works? i.e. I don't think it necessarily works in serial like a pipeline from front buffer > 2nd buffer > 3rd buffer. Rather, the 2nd and 3rd buffer are in parallel, feeding into the front buffer. Whichever one has the latest content is the one which will be fed into the front buffer. So if the drawing code is running faster than the screen refresh rate, often the content in the second buffer will be skipped entirely, in favour of the later-drawn third buffer. The point of triple buffering is to _avoid_ lags in getting the latest content to the screen. So I'm not sure that this sort of code could actually show any difference between double and triple buffering. 
If we are seeing a one-frame lag, I suspect that might be due to something happening at the OS level (i.e. like the extra processing sometimes done in monitor hardware), i.e. a level of buffering which is laid on top of our user-level drawing code. As per @peircej's suggestion, I don't think we can make much headway here without external hardware timing validation.

**Comment by hoechenberger** [comment]
> Does this reflect how triple buffering actually works? 

A problem is that "triple buffering" seems to be somewhat of an umbrella term with slightly different meanings, but my understanding is that, yes, this is generally how it's supposed to work.

> i.e. I don't think it necessarily works in serial like a pipeline from 3rd buffer -> 2nd buffer -> front buffer. Rather, the 2nd and 3rd buffer are in parallel, feeding into the front buffer. Whichever one has the latest content is the one which will be fed into the front buffer.

The question is: How would the driver know which buffer has the "latest" content? By issuing the `flip` command, you essentially mark a drawing operation as done and a buffer as "ready to be displayed"; then you may immediately start filling the second buffer, issue `flip`, immediately start filling the first buffer again, etc.

In that sense, unless there's some kind of interaction on an additional level (OS, hardware), I believe the test I presented would be valid.

At any rate, I will try to hook up a BBTK some time next week to my computer, to test with both internal and external displays. A colleague of mine has one, and I'll ask if he'd be willing to give it to me for a couple hours.

**Comment by peircej** [comment]
Yes, I don't think this test would *necessarily* show the problem because a) triple buffering might be implemented in multiple ways and b) we don't even know for sure that it's the problem. If we find a software solution that demonstrates the problem it will be great to help us, but if it doesn't show the problem then it only suggests that the test isn't sensitive!
@dvbridges can pass on the code. We bought the "key actuator" (robotic keyboard basher) for the BBTK and that allows us to test the reaction time delay too, but for this test all you need is a labjack or equivalent to send a trigger pulse at (apparent) screen flip.

**Comment by hoechenberger** [comment]
> @dvbridges can pass on the code. We bought the "key actuator" (robotic keyboard basher) for the BBTK and that allows us to test the reaction time delay too, but for this test all you need is a labjack or equivalent to send a trigger pulse at (apparent) screen flip.

Cool! I have an NI DAQ board, just need to find out if it works with my Mac :)

**Comment by dvbridges** [comment]
@hoechenberger , here is the code. I use ```Digital Stimulus Capture``` for collecting timing data, and ```Digital Stimulus Response Echo``` for response time data. They are run separately because there is a limited number of trials that can be run with our BBTK, but with these two methods we can run 1000 trials without any problems. The BBTK module is in the PsychoPy code already :) 

[runBBTK.zip](https://github.com/psychopy/psychopy/files/2827632/runBBTK.zip)

**Comment by hoechenberger** [comment]
Great, thank you!

**Comment by cbrnr** [comment]
Not sure if this is related, but I have severe issues with visual timing on my MacBook Pro (Mid-2014) running Mojave 10.14.3 (PsychoPy 3.0.4): `win.getActualFrameRate()` gives me values around 430 (with pretty high variance). I've tried this with the internal display as well as with an external one (both run at 60Hz), but I get these values for both internal and external displays. I've also tried the `pyglet` and `glfw` backends, but the problem seems to be independent of the backend.

**Comment by peircej** [comment]
No, I think that's unrelated. The issue I mention is about an offset of 1 frame but the sync to refresh is stable. 
Your machine says the frame rate is 430?! That's something pretty broken, that I've never seen on a mac. On windows this would indicate that you weren't synching to the screen refresh but on a mac I've never had that problem.
Could you upload the graph from timeByFrames (in the demo menu) here?

**Comment by cbrnr** [comment]
Here's the graph:
![figure_1](https://user-images.githubusercontent.com/4377312/53407342-483b6e00-39bc-11e9-905d-2c1b23ea7e7e.png)

**Comment by oschmidtmann** [comment]
@cbrnr Your frame times should variate around 1/60 (16.67) ms. Your frame times are around 50. This is very odd. One important factor for MacBooks for me was to use a single display only. Otherwise the fps deviations got higher. For testing I'd suggest to physically detach the external one.  

But there seems to be another issue. Have you tried to build a very simple test with just one shape being drawn one time before win.flip? This is a good point to start. Especially core.wait (including waitkeys!) may cause a high load.

Also: the spike around frame 450 may be related to your programming – which is hard to say. It might also be worthwhile to shut down all possible processes (like timemachine, spotlight indexing, any cloud syncs etc.). If this does not help you might want introduce certain points in your code  at which you print the current frame and the time to find the bottleneck.
Good luck!

**Comment by cbrnr** [comment]
Yes, with the internal display I do get the expected output graph using the `pyglet` backend but not the `glfw` backend. However, in all cases (even in the case with only using the internal display and `pyglet`) I do get strangely high values around 400 from `win.getActualFrameRate()`.

@peircej should I open a separate issue?

**Comment by peircej** [comment]
I'm confused what you refer to when you say 450. 

- I thought you meant the frame rate (450 frames per second) which would have meant that you weren't syncing to the flip. 
- In fact your frame times are roughly 50ms, which indicates the opposite problem: frames are being dropped pretty consistently. Were you meaning 450 ms?
- But in the graph on the left there's a single frame that's taking a very long time at frame number 450 or so. Is that what you were talking about. That indicates some other issue, that the computer is suddenly spending time doing something or other (like synchronising with dropbox, or checking the weather!....)

If everything looks normal without the external monitor connected but does this when you connect a display then I think it's something to do with you laptop trying to synchronise two displays that might not even have the same native frequency). If you consistently get this spike around frame 450 it might be taking that long for the machine to detected the fullscreen mode and then does does something like re-sync with the 2 displays. There are lots of variables here (like the mode of screens, mirroring or not) and how they are set up (optimized for external etc).

You might also try just turning off the window waitblanking setting because that might be interacting badly with your operating system trying to perform this 2-screen sync.

But critically, could you try and always give the _unit_ when you mention a number, because it dramatically changes what the number means! ;-)

**Comment by cbrnr** [comment]
OK, sorry for the confusion. There are a lot of parameters here that I didn't explain clearly, so maybe it's best to present them in a table. The last column shows the output of `win.getActualFrameRate()`, which is the frame rate in Hz. I use PsychoPy 3.0.4 running on Python 3.7.2.

### macOS 10.14.3

Here are the results if I run the visualization on my external display:

winType | fullscr | waitBlanking | getActualFrameRate()
------- | ------- | ------------ | --------------------
pyglet  | True    | True         |                 None
pyglet  | True    | False        |                 None
pyglet  | False   | True         |               380.07
pyglet  | False   | False        |               511.33
glfw    | True    | True         |               455.39
glfw    | True    | False        |               539.14
glfw    | False   | True         |               401.31
glfw    | False   | False        |               508.77

And here's how these values look like on my internal display:

winType | fullscr | waitBlanking | getActualFrameRate()
------- | ------- | ------------ | --------------------
pyglet  | True    | True         |                59.85
pyglet  | True    | False        |                 None
pyglet  | False   | True         |               440.07
pyglet  | False   | False        |               639.94
glfw    | True    | True         |                90.95
glfw    | True    | False        |                 None
glfw    | False   | True         |               473.26
glfw    | False   | False        |               623.38

Obviously, I get the expected frame rate of 60Hz only on my internal display using `winType="pyglet"`, `fullscr=True`, and `waitBlanking=True`.

### Arch Linux
I only have one display on that machine (Xorg, proprietary `nvidia` driver):

winType | fullscr | waitBlanking | getActualFrameRate()
------- | ------- | ------------ | --------------------
pyglet  | True    | True         |                29.98
pyglet  | True    | False        |                59.98
pyglet  | False   | True         |                29.97
pyglet  | False   | False        |                59.98
glfw    | True    | True         |                60.31
glfw    | True    | False        |                60.35
glfw    | False   | True         |                60.33
glfw    | False   | False        |                60.34

On Linux, all combinations seem to work except for the (known) issue that I only get half the frame rate if I set `waitBlanking=True` in combination with `winType="pyglet"` (it would be super nice if this issue was resolved, because currently the only configuration that works on macOS does not work on Linux). Plus, `waitBlanking` doesn't have the desired effect since vsync seems to be always enabled no matter what.

### Script
Here's the script to produce this table in case anyone wants to try it on their machine:

```python
from psychopy import core, visual


wintypes = ["pyglet", "glfw"]
fullscreen = [True, False]
vsync = [True, False]

table = ("winType | fullscr | waitBlanking | getActualFrameRate()\n"
         "------- | ------- | ------------ | --------------------\n")

for wintype in wintypes:
    for fullscr in fullscreen:
        for waitblank in vsync:
            table += f"{wintype:7} | {str(fullscr):7} | {str(waitblank):12} | "
            win = visual.Window(winType=wintype, fullscr=fullscr,
                                waitBlanking=waitblank)
            framerate = win.getActualFrameRate()
            win.close()
            if framerate is None:
                table += f"{'None':>20}\n"
            else:
                table += f"{framerate:20.2f}\n"
print()
print(table)
core.quit()
```

**Comment by cbrnr** [comment]
I ran the script on our Windows machine with a 120Hz monitor, here's the output:

winType | fullscr | waitBlanking | getActualFrameRate()
------- | ------- | ------------ | --------------------
pyglet  | True    | True         |               119.99
pyglet  | True    | False        |               119.99
pyglet  | False   | True         |               120.21
pyglet  | False   | False        |               119.87
glfw    | True    | True         |                60.32
glfw    | True    | False        |                60.34
glfw    | False   | True         |               118.14
glfw    | False   | False        |               120.92

On this platform, only `glfw` doesn't seem to work correctly in fullscreen mode.

**Comment by isolver** [comment]
Re the extra 1 frame delay on newer versions of macOS, I've have also seen this on macOS 10.13.6 testing with the MilliKey light sensor. The exact same script on the same computer  running in a 10.9.5 partition shows no extra frame delay. I think @peircej also reproduced this in his lab? 

From what I gather, macOS < 10.13 is fine, but anything >= 10.13 has a 1 frame delay when tested with a light key.

**Comment by mdcutone** [comment]
@cbrnr 
> On this platform, only `glfw` doesn't seem to work correctly in fullscreen mode.

Try using the argument 'swapInterval=1' when using GLFW in fullscreen mode. GLFW has multiple v-sync settings and I think waitBlanking might not be setting it correctly.

**Comment by cbrnr** [comment]
@mdcutone unfortunately, this didn't seem to have any effect. It's still showing as 60Hz.

**Comment by nahorp** [comment]
> Plus, `waitBlanking` doesn't have the desired effect since vsync seems to be always enabled no matter what.

I am experiencing the same as well @cbrnr. Any update regarding `waitBlanking` not having the desired effect?

**Comment by peircej** [comment]
*Sync* to the blank and *wait* for the vertical blank are actually two distinct concepts. Setting `waitBlanking=False` does not turn off sync and I don't intend to change that. See https://www.psychopy.org/general/rendering.html#sync-to-vbl-and-wait-for-vbl

**Comment by pellet** [comment]
I was unable to get over 85Hz on my external display on macOS 11.5.2(Big Sur) without it returning None instead of the correct frame rate.  I just dropped the refresh rate as a work around.

**Comment by waltervanheuven** [comment]
I'm unable to get correct frame rate detection with high refresh monitors on PC (main screen) and Mac (using external monitor) with the latest versions of PsychoPy. I'm using frame-based timings in the Builder and I can only get accurate timing when setting monitors to 60Hz!

Results below when running the script that was posted above (adapted to test also external screen).

Setup

- M1 MacBook Pro
- macOS Montery 12.2.1
- External Monitor: BenQ EX2780Q
- Extended display for dual monitor setup

## PsychoPy v2021.2.3

Only `pyglet` backend for windows is working. Note that 2nd screen never shows a grey full screen.

### Setup 1

screen 0: internal 2560x1600 60Hz   (mainScreen - retina)
screen 1: external 2560x1440 144Hz

| Screen | winType | fullscr | waitBlanking | getActualFrameRate() |
| ------ | ------- | ------- | ------------ | -------------------- |
|      0 | pyglet  | True    | True         |                59.43 |
|      0 | pyglet  | True    | False        |                60.01 |
|      0 | pyglet  | False   | True         |                59.76 |
|      0 | pyglet  | False   | False        |                59.44 |
|      1 | pyglet  | True    | True         |                59.93 |
|      1 | pyglet  | True    | False        |                59.97 |
|      1 | pyglet  | False   | True         |                 None |
|      1 | pyglet  | False   | False        |                 None |

### Setup 2

screen 0: internal 2560x1600 60Hz   (mainScreen - retina)
screen 1: external 2560x1440 120Hz
      
|Screen | winType | fullscr | waitBlanking | getActualFrameRate() |
| ------ | ------- | ------- | ------------ | -------------------- |
|      0 | pyglet  | True    | True         |                60.07 |
|      0 | pyglet  | True    | False        |                60.11 |
|      0 | pyglet  | False   | True         |                60.03 |
|      0 | pyglet  | False   | False        |                59.86 |
|      1 | pyglet  | True    | True         |                60.20 |
|      1 | pyglet  | True    | False        |                59.88 |
|      1 | pyglet  | False   | True         |                59.49 |
|      1 | pyglet  | False   | False        |                59.50 |
    
### Setup 3

screen 0: external 2560x1600 144Hz	(mainScreen)
screen 1: internal 5260x2880 60Hz   
      
| Screen | winType | fullscr | waitBlanking | getActualFrameRate() |
| ------ | ------- | ------- | ------------ | -------------------- |
|      0 | pyglet  | True    | True         |                 None |
|      0 | pyglet  | True    | False        |                 None |
|      0 | pyglet  | False   | True         |                 None |
|      0 | pyglet  | False   | False        |                72.47 |
|      1 | pyglet  | True    | True         |              1415.23 |
|      1 | pyglet  | True    | False        |              1262.10 |
|      1 | pyglet  | False   | True         |                59.79 |
|      1 | pyglet  | False   | False        |                59.96 |


### Setup 4

screen 0: external 2560x1600 120Hz	(mainScreen)
screen 1: internal 5260x2880 60Hz   

| Screen | winType | fullscr | waitBlanking | getActualFrameRate() |
| ------ | ------- | ------- | ------------ | -------------------- |
|      0 | pyglet  | True    | True         |                60.14 |
|      0 | pyglet  | True    | False        |                60.05 |
|      0 | pyglet  | False   | True         |                59.82 |
|      0 | pyglet  | False   | False        |                59.97 |
|      1 | pyglet  | True    | True         |              1443.63 |
|      1 | pyglet  | True    | False        |              2591.37 |
|      1 | pyglet  | False   | True         |                59.76 |
|      1 | pyglet  | False   | False        |                59.77 |


## PsychoPy v2022.1.1

The `pygame` backend for windows is not working. Note that 2nd screen never shows a grey full screen.

### Setup 1

screen 0: internal 2560x1600 60Hz   (mainScreen - retina)
screen 1: external 2560x1440 144Hz

|Screen | winType | fullscr | waitBlanking | getActualFrameRate() |
| ------ | ------- | ------- | ------------ | -------------------- |
|      0 | pyglet  | True    | True         |                60.73 |
|      0 | pyglet  | True    | False        |                60.00 |
|      0 | pyglet  | False   | True         |                59.64 |
|      0 | pyglet  | False   | False        |                60.78 |
|      0 | glfw    | True    | True         |                59.91 |
|      0 | glfw    | True    | False        |              1030.72 |
|      0 | glfw    | False   | True         |                59.55 |
|      0 | glfw    | False   | False        |              2346.34 |
|      1 | pyglet  | True    | True         |               734.24 |
|      1 | pyglet  | True    | False        |                60.90 |
|      1 | pyglet  | False   | True         |                71.15 |
|      1 | pyglet  | False   | False        |                 None |
|      1 | glfw    | True    | True         |                59.00 |
|      1 | glfw    | True    | False        |              1095.51 |
|      1 | glfw    | False   | True         |                 None |
|      1 | glfw    | False   | False        |              3102.94 |

### Setup 2

screen 0: internal 2560x1600 60Hz   (mainScreen - retina)
screen 1: external 2560x1440 120Hz

|Screen | winType | fullscr | waitBlanking | getActualFrameRate() |
| ------ | ------- | ------- | ------------ | -------------------- |
|      0 | pyglet  | True    | True         |                60.23 |
|      0 | pyglet  | True    | False        |                59.97 |
|      0 | pyglet  | False   | True         |                59.67 |
|      0 | pyglet  | False   | False        |                60.07 |
|      0 | glfw    | True    | True         |                60.10 |
|      0 | glfw    | True    | False        |              1046.60 |
|      0 | glfw    | False   | True         |                60.35 |
|      0 | glfw    | False   | False        |              2309.05 |
|      1 | pyglet  | True    | True         |               784.94 |
|      1 | pyglet  | True    | False        |              1039.20 |
|      1 | pyglet  | False   | True         |                59.80 |
|      1 | pyglet  | False   | False        |                59.50 |
|      1 | glfw    | True    | True         |                59.70 |
|      1 | glfw    | True    | False        |              1012.12 |
|      1 | glfw    | False   | True         |                60.53 |
|      1 | glfw    | False   | False        |              3170.95 |

**Comment by peircej** [comment]
While I agree this would be good to fix, I'm going to reiterate that I wouldn't recommend using MacOS for running actual experiments that need precise frame timing (with any software package) because of the poor frame onset timing that nobody has been able to fix on MacOS for many versions of the operating system. See our timing paper for more details

**Comment by peircej** [comment]
(I guess I'm saying that it might be MacOS has introduced some new fun option that means the timing doesn't work)
